flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/rest/util/RowDataLocalTimeZoneConverter.java

```diff
@@ -0,0 +1,171 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.gateway.rest.util;
+
+import org.apache.flink.configuration.ReadableConfig;
+import org.apache.flink.table.api.config.TableConfigOptions;
+import org.apache.flink.table.data.ArrayData;
+import org.apache.flink.table.data.GenericArrayData;
+import org.apache.flink.table.data.GenericMapData;
+import org.apache.flink.table.data.GenericRowData;
+import org.apache.flink.table.data.MapData;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.data.TimestampData;
+import org.apache.flink.table.types.logical.ArrayType;
+import org.apache.flink.table.types.logical.LocalZonedTimestampType;
+import org.apache.flink.table.types.logical.LogicalType;
+import org.apache.flink.table.types.logical.LogicalTypeRoot;
+import org.apache.flink.table.types.logical.MapType;
+import org.apache.flink.table.utils.DateTimeUtils;
+
+import java.time.ZoneId;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.TimeZone;
+
+/**
+ * Convert {@link LocalZonedTimestampType} data in {@link RowData} to {@link TimestampData} based on
+ * the local time zone.
+ */
+public class RowDataLocalTimeZoneConverter {
+    private final List<RowData.FieldGetter> fieldGetterList;
+    private final List<LogicalType> logicalTypeList;
+    private final boolean hasTimeZoneData;
+    private final TimeZone timeZone;
+
+    public RowDataLocalTimeZoneConverter(List<LogicalType> logicalTypeList, ReadableConfig config) {
+        this(logicalTypeList, TimeZone.getTimeZone(getSessionTimeZone(config)));
+    }
+
+    public RowDataLocalTimeZoneConverter(List<LogicalType> logicalTypeList, TimeZone timeZone) {
+        this.logicalTypeList = logicalTypeList;
+        this.timeZone = timeZone;
+        this.fieldGetterList = new ArrayList<>(logicalTypeList.size());
+        for (int i = 0; i < logicalTypeList.size(); i++) {
+            fieldGetterList.add(RowData.createFieldGetter(logicalTypeList.get(i), i));
+        }
+        this.hasTimeZoneData = checkTimeZoneType(logicalTypeList);
+    }
+
+    private boolean checkTimeZoneType(List<LogicalType> logicalTypeList) {
+        for (LogicalType logicalType : logicalTypeList) {
+            if (logicalType.getTypeRoot() == LogicalTypeRoot.TIMESTAMP_WITH_LOCAL_TIME_ZONE) {
+                return true;
+            } else if (logicalType.getTypeRoot() == LogicalTypeRoot.MAP) {
+                MapType mapType = (MapType) logicalType;
+                if (checkTimeZoneType(
+                        Arrays.asList(mapType.getKeyType(), mapType.getValueType()))) {
+                    return true;
+                }
+            } else if (logicalType.getTypeRoot() == LogicalTypeRoot.ARRAY) {
```

Comment:
There are also various type roots such as `MULTISET`, `ROW`, do you think we also need to handle them too?

Reply:
👍 Added MULTISET and ROW